controllers/userController.ts

```diff
@@ -0,0 +1,60 @@
+//\\//\\//\\//\\//\\//\\//\\//\\//\\//\\//\\
+// USER ENTITY UPDATE & DELETE FUNCTIONS  \\
+//\\//\\//\\//\\//\\//\\//\\//\\//\\//\\//\\
+
+import { RequestHandler, Request, Response, NextFunction } from "express";
+import bcrypt from "bcrypt";
+
+// Import TS types
+import { UserMethods } from "../models/userModel";
+import { User } from "@prisma/client";
+
+// Extend express-session SessionData to include user data
+declare module "express-session" {
+    interface SessionData {
+        user: User;
+        loggedIn: boolean;
+    }
+}
+
+// Update controller
+// -------------------------------------------------------------------------
+
+export const update =
+    (
+        getUserByEmail: UserMethods.GetUserByEmail,
+        getUserByUserName: UserMethods.GetUserByUserName,
+        updateUser: UserMethods.UpdateUser
+    ): RequestHandler =>
+    async (
+        req: Request,
+        res: Response,
+        next: NextFunction
+    ): Promise<void | Response> => {
+        try {
+            // Check if username and email are already taken
+            const userNameExists = await getUserByUserName(req?.body?.userName);
+            const emailExists = await getUserByEmail(req?.body?.email);
+            if (
+                (userNameExists &&
+                    req?.session?.user?.id !== userNameExists?.id) ||
+                (emailExists && req?.session?.user?.id !== emailExists?.id)
+            )
+                return res.status(400).json({
+                    status: "error",
+                    message:
+                        "A user already ``   exists with that username or email address",
+                });
+
+            // Encrypt the password
+            const password = req.body.password;
```

Comment:
I strongly recommend removing password from this update route for a number of reasons. For security, the password is removed before being sent to the frontend in requests, this means that forms won't pre-fill the password --> if validation makes passwords optional here then we could have a null value pushed to the db as password --> the user no longer has a password. If validation requires the password, then a user will need to enter their password to update any other information. It's also more secure to have a dedicated password reset system, we could then integrate 2FA or email validation or whatever else in future as well. This would also remove the possibility of a password accidentally being changed

Reply:
Seems that this comment is still in the code 😃

Reply:
Good eyes! I'm not sure what happened there, maybe I didn't save all changes before the last commit but it should be fixed now.

Reply:
No probs